x-pack/plugins/ml/common/util/fields_utils.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  Field,
+  Aggregation,
+  NewJobCaps,
+  METRIC_AGG_TYPE,
+  RollupFields,
+  EVENT_RATE_FIELD_ID,
+} from '../types/fields';
+import { ES_FIELD_TYPES } from '../../../../../src/plugins/data/common';
+import { ML_JOB_AGGREGATION } from '../constants/aggregation_types';
+
+// cross reference fields and aggs.
+// fields contain a list of aggs that are compatible, and vice versa.
+export function combineFieldsAndAggs(
+  fields: Field[],
+  aggs: Aggregation[],
+  rollupFields: RollupFields
+): NewJobCaps {
+  const keywordFields = getKeywordFields(fields);
+  const textFields = getTextFields(fields);
+  const numericalFields = getNumericalFields(fields);
+  const ipFields = getIpFields(fields);
+  const geoFields = getGeoFields(fields);
+
+  const isRollup = Object.keys(rollupFields).length > 0;
+  const mix = mixFactory(isRollup, rollupFields);
+
+  aggs.forEach((a) => {
```

Comment:
All other code in this file is creating new instances of arrays so I wonder if the code here would be more clear if it was using `.map()` and reassigned to a new `aggs` instances, what do you think? For me it was a bit hard to follow the code because it's mutating the aggregations. As an alternative maybe add a comment here that this is mutating `aggs` with the inner `forEach`s?

Reply:
i'm hesitant to change this code at all in this PR as it is not new, I have just moved it from the server side into common so it can be used on the client side.